text/0000-variant-types.md

```diff
@@ -0,0 +1,335 @@
+operations on the interior data. However, in many cases there is a large amount
+of processing to be done. Ideally we would factor that out into a function,
+passing the data to the function. However, currently in Rust, enum variants are
+not types and so we must choose an unsatisfactory work around - we pass
+each field of the variant separately (leading to unwieldy function signatures
+and poor maintainability), we pass the whole variant with enum type (and have to
+match again, with `unreachable!` arms in the function), or we embed a struct
+within the variant and pass the struct (duplicating data structures for no good
+reason). It would be much nicer if we could refer to the variant directly in the
+type system.
+
+When working with FFI code, we need to communicate with C programs which may use
+union data types. There is no way to represent a union in Rust, and thus working
+with such types is awkward and involves bug-prone transmutes. We should provide
+some way for Rust to handle such types.
+
+As we'll see below, variant types allow for an elegant solution to the union
+problem.
+
+
+# Detailed design - variant types
+
+Consider the example enum `Foo`:
+
+```rust
+pub enum Foo {
+    Variant1,
+    Variant2(i32, &'static str),
+    Variant3 { f1: i32, f2: &'static str },
+}
+```
+
+We create new instances by constructing one of the variants. The only type
+introduced is `Foo`. Variant names can only be used in patterns and for creating
+instances. E.g.,
+
+```rust
+fn new_foo() -> Foo {
+    Foo::Variant2(42, "Hello!")
+}
+```
+
+This RFC proposes allowing the programmer to use variant names as types, e.g.,
+
+```rust
+fn bar(x: Foo::Variant2) {}
+struct Baz {
+    field: Foo::Variant3,
+}
+```
+
+Both enums and their variants can currently be imported:
+
+```rust
+use Foo;
+use Foo::Variant1;
+```
+
+Importing an enum imports it into both the value and type namespace. Importing
+a variant imports it only into the value namespace. To maintain backwards
+compatibility, this will remain the default. In order to import an enum variant
+into the type namespace, one must use the `import_variant_type` attribute:
+
+```rust
+use Foo;
+#[import_variant_type]
+use Foo::Variant1;
+
+fn bar(v: Variant1) {
+    let _ = Variant1;
+}
+```
+
+When we release Rust v2.0, we may choose to import variants into both namespaces
+by default and remove the attribute.
+
+
+## Constructors
+
+Consider `let x = Foo::Variant1;`, currently `x` has type `Foo`. In order to
+preserve backwards compatibility, this must remain the case. However, it would
+be convenient for `let x: Foo::Variant1 = Foo::Variant1;` to also be valid.
+
+The type checker must consider multiple types for an enum construction
+expression - both the variant type and the enum type. If there is no further
+information to infer one or the other type, then the type checker uses the enum
+type by default. This is analogous to the system we use for integer fallback or
+default type parameters.
```

Comment:
This is great, pretty much what I expected and @nikomatsakis also confirmed it's what he would do (although him and @aturon aren't sure it's enough i.e. compared to full subtyping).